R/layers2traces.R

```diff
@@ -55,16 +56,20 @@ layers2traces <- function(data, prestats_data, layout, p) {
     trs <- Map(geom2trace, dl, paramz[i], list(p))
     # are we splitting by a discrete scale on this layer?
     # if so, set name/legendgroup/showlegend
-    isDiscrete <- names(d) %in% paste0(names(discreteScales), "_plotlyDomain")
+    isDiscrete <- names(d) %in% (if (length(discreteScales) > 0) { paste0(names(discreteScales), "_plotlyDomain") } else as.character())
     if (length(trs) > 1 && sum(isDiscrete) >= 1) {
-      nms <- names(trs)
-      # ignore "non-discrete" scales that we've split on
-      for (w in seq_len(sum(names(d) %in% c("PANEL", split_on(d))))) {
-        nms <- sub("^[^@%&]@%&", "", nms)
-      }
-      nms <- strsplit(nms, "@%&")
+      nms <- strsplit(names(trs), "@%&", fixed = TRUE)
+      # ignore "non-discrete" scales that we've split by
+      n_ignore <- sum(!isDiscrete & (names(d) %in% split_by))
       nms <- vapply(nms, function(x) {
-        if (length(x) > 1) paste0("(", paste0(x, collapse = ","), ")") else x
+        n_x <- length(x)
+        if (n_x > n_ignore + 1) {
+          paste0("(", paste0(x[(n_ignore+1):n_x], collapse = ","), ")")
+        } else if (n_x > n_ignore) {
+          x[[n_ignore+1]]
+        } else {
+          NA_character_
+        }
       }, character(1))
```

Comment:
Instead of removing the first ignored scales using `sub()` in the loop, I've changed the code to directly split the trace ids by "@%&" and then ignore the first `n_ignore` chunks (non-discrete scales). Then the visible trace names are composed as before: if there's more than one scale left after `n_ignore`, the trace name is like `(a, b)`, otherwise the remaining scale name is used "as is".
In the original code there was an error in the `sub()` regex that manifested itself when the length of the non-discrete scale id was more than 1 symbol. So I thought it would be cleaner and safer to avoid preprocessing trace Ids.